Topics/Tech_Stacks/Spring-Security.md

```diff
@@ -0,0 +1,104 @@
+# REST API Security
+- Spring Security Model
+	- Spring Security defines a framework for security
+	- Implemented using Servlet filters in the background
+		- Servlet filters pre-process / post-process web requests
+		- Servlet filters route web requests based on security logic
+		- Spring provides a bulk of security functionality with servlet filters
+	- Two methods of securing an app: declarative and programmatic
+
+### Security Concepts
+- Authentication
+	- Check user id and password with credentials stored in app / db
+- Authorization
+	- Check to see if user has an authorized role
+- Declarative Security
+	- Define application's security constraints in configuration
+		- All Java config: @Configuration
+	- Provides separation of concerns between application code and security
+- Programmatic Security
+	- Spring Security provides an API for custom application coding
+	- Provides greater customization for specific app requirements
+
+### Enabling Spring Security
+1. Edit pom.xml and add `spring-boot-starter-security`
+2. This will automagically secure all endpoints for application
+	- Accessing secured endpoints will prompt for login
+	- Override default user name and generated password in application.properties
+
+- Cross-Site Request Forgery (CSRF)
+	- Spring Security can protect against CSRF attackes
+	- Embed additional authentication data / token into all HTML forms
+	- On subsequent requests, web app will verify token before processing
+	- Primary use case is traditional web applications (HTML forms etc...)
+
+- When to use CSRF Protection
+	- For any normal browser web requests
+	- Traditional web apps with HTML forms to add / modify data
+	- If building REST API for non-browser clients
+		- disable CSRF protection, not required for stateless REST APIs that use POST, PUT, DELETE and / or PATCH
+
+### Database Access
+- Default Spring Security Database Schema
+	- users
+		- username
+		- password
+		- enabled
+	- authorities (same as roles)
+		- username
+		- authority
+
+- BCrypt Encryption
+	- Performs one-way encrypted hashing
+	- Adds a random salt to the password for additional protection
+	- Includes support to defeat brute force attack
+
+	- Generate password
+		- www.luv2code.com/generate-bycrypt-password
```

Comment:
This link reports 404, please consider removing it or replace it.

Reply:
Regarding the broken link, I appreciate you pointing that out. I'll make sure to remove or replace it with a functional and relevant one.